python/src/aiconfig/eval/README.md

```diff
@@ -0,0 +1,110 @@
+
+Inputs:
+
+```
+test_inputs = [
+    "Food",
+    "Iconic midtown skyscrapers",
+]
+expected_substrings = [
+    "Magnolia Bakery",
+    "Empire State Building",
+]
+
+# For each input, ask the library to check for substring match and brevity (string length in chars).
+...
+```
+
+![Example AIConfig Eval output table](image-4.png)
+
+# Usage Guide - Custom Eval
+
+
+## Quick start:
+
+Please see [this notebook](https://github.com/lastmile-ai/aiconfig/blob/d79cf7dadf934b0ce09bc671cfde37aaecf05c1e/python/src/aiconfig/eval/custom_eval/examples/travel/travel_eval.ipynb#L4) for an example of running evaluation on an existing AIConfig. Make sure the AIConfig library is installed and then run the notebook cells in order.
+
+
+## Custom Eval Usage Steps
+
+Let's take a closer look at what we're testing and how. The travel example linked above is a parametrized AIConfig with the following DAG topology (it is linear in this example):
+
+`the_query -> get_activities.input -> get_activities.output -> gen_itinerary.input -> gen_itinerary.output`. 
+
+The leaf "the_query" is the placeholder for test case inputs, like a function argument. The root "gen_itinerary.output" is the value under test. You must specify exactly one leaf input and it must be named "the_query". You must also specify the name of the prompt you want to test (see test suite settings in the example notebook). 
+
+
+Zooming in on the edges, each of which is either prompt string formatting or model execution:
+
+```
+the_query -[string formatting]-> 
+  get_activities.input -[get_activities]-> 
+  get_activities.output -[string formatting]-> 
+  gen_itinerary.input -[gen_itinerary]-> 
+  gen_itinerary.output
+  ```
+  .
+
+
+## Evaluate Precomputed Outputs
+If you have already run your AIConfig and saved its outputs (either in memory or a data store) you can use the library to evaluate that dataset similar to the way described above. The idea is to define your test cases as 
+
+`(output, evaluation_metric)` 
+
+instead of 
+
+`(input, evaluation_metric)`. 
+
+The library will just run your evaluations, skipping the AIConfig inference step. Due to this difference, you do not need to define test suite settings in this case.
+
+See the bottom half of the notebook linked above for a usage example.
+
+
+# Usage Guide - Promptfoo Integration
+
+If you are already familiar with [Promptfoo](https://www.promptfoo.dev/) and/or prefer to leverage it for your AIConfig evaluation, please follow the steps below. **<u>Some familiarity with Promptfoo is assumed</u>**. Please refer to their documentation as needed.
+
+## Quick start
+
+Please run the example as follows:
+
+`npx promptfoo@latest eval -c python/src/aiconfig/eval/promptfoo/examples/travel/travel_promptfooconfig.yaml`
+
+You should get one passing test and one failing test:
+
+![Example Promptfoo AIConfig Integration Output](image-5.png)
+
```

Comment:
Is it possible to add a cookbook for this? If not I think it would also be good to have another separate <PromptFooDetails.md> file. I know I gave feedback yesterday to make it all in one place instead of all over, but I feel that now that we have a quickstart to quickly show what it does and how it's useful, think having more detailed other than README would be easier to consume!

Reply:
I put detail below, and good question about cookbook. I'm not sure exactly what that would entail beyond the configs. There is no custom code needed for promptfoo integration.